### PR TITLE
New channels

### PR DESCRIPTION
**Pull request by josh146**
**Description of the Change:**

* Thermal loss channel is now implemented in the Gaussian backend
* Added `thermal_loss()` to `BaseBackend`. At the moment, since it is not implemented in the Fock backends, attempting to use it there will raise a `NotImplementedError`.
* Added `ThermalLossChannel` operation to `ops.py`, and moved the common channel merge method up to `Channel`.
* Added backend tests for `thermal_loss` to the `test_loss_channel.py` file.
* Added documentation for the thermal loss channel to `doc/conventions/others.rst`.

**Benefits:**

* The addition of another loss channel.

**Possible Drawbacks:**

* Currently not implemented in the Fock backends - a Kraus operator mapping will need to be implemented.

**Related GitHub Issues:**

* n/a

### CODE REVIEW

File: doc/conventions/others.rst [hunk 1]
```diff
@@ -48,10 +48,10 @@ Loss is implemented by a CPTP map whose Kraus representation is
 .. admonition:: Definition
     :class: defn
 
-    Loss is implemented by coupling mode :math:`\a` to another bosonic mode prepared in the vacuum state, by using the following transformation
+    Loss is implemented by coupling mode :math:`\a` to another bosonic mode :math:`\hat{b}` prepared in the vacuum state, by using the following transformation
```
Comment (co9olguy) [inline doc/conventions/others.rst]: add hat to a

File: doc/conventions/others.rst [hunk 2]
```diff
@@ -68,6 +68,28 @@ One useful identity is
...
 
+.. _thermal_loss:
+
+Thermal loss channel
+---------------------------------------------
+
+.. admonition:: Definition
+    :class: defn
+
+    Thermal loss is implemented by coupling mode :math:`\a` to another bosonic mode :math:`\hat{b}` prepared in the thermal state :math:`\ket{\bar{n}}`, by using the following transformation
```
Comment (co9olguy) [inline doc/conventions/others.rst]: same

File: doc/conventions/others.rst [hunk 3]
```diff
@@ -68,6 +68,28 @@ One useful identity is
...
+
+    .. math::
+       \a \to \sqrt{T} \a+\sqrt{1-T} \hat{b}
+
+    and then tracing it out. Here, :math:`T` is the *energy* transmissivity. For :math:`T = 0` the state is mapped to the thermal state :math:`\ket{\bar{n}}`, and for :math:`T=1` one has the identity map.
```
Comment (co9olguy) [inline doc/conventions/others.rst]: --> mapped to the thermal state with mean photon number nbar

File: strawberryfields/ops.py [hunk 4]
```diff
@@ -616,10 +617,28 @@ class Channel(Transformation):
     This class provides the base behaviour for non-unitary
     maps and transformations.
     """
-    # NOTE: At the moment this is an empty class, and only
-    # exists for a nicer inheritence diagram.
-    # This class will likely get filled in when we add more channels.
-    pass
+    def merge(self, other):
+        # check that other an identical channel, and that
```
Comment (co9olguy) [inline strawberryfields/ops.py]: typo. "is an identical channel"?

File: strawberryfields/ops.py [hunk 5]
```diff
@@ -616,10 +617,28 @@ class Channel(Transformation):
...
+        # no extra dependencies <=> no RegRefTransform parameters
```
Comment (co9olguy) [inline strawberryfields/ops.py]: "that there are no extra dependencies"?

File: tests/test_loss_channel.py [hunk 6]
```diff
@@ -20,97 +20,167 @@
...
-    ttt = unittest.TestLoader().loadTestsFromTestCase(t)
-    suite.addTests(ttt)
+    # run the tests in this file
+    suite = unittest.TestSuite()
+    for t in (BasicTests, GaussianBasisTests, FockBasisTests):
+        ttt = unittest.TestLoader().loadTestsFromTestCase(t)
+        suite.addTests(ttt)
 
-  unittest.TextTestRunner().run(suite)
+    unittest.TextTestRunner().run(suite)
```
Comment (co9olguy) [inline tests/test_loss_channel.py]: Nice tests! :)

**Comment by co9olguy** [comment]
Note that I did not review the actually backend changes here (those were ninja'd in directly to master)

**Comment by co9olguy** [comment]
What's the story with the failing coverage report?

File: strawberryfields/backends/gaussianbackend/gaussiancircuit.py [hunk 7]
```diff
@@ -357,15 +357,15 @@ def loss(self, T, k):
         self.mmat[:, k] = self.mmat[k]
         self.mean[k] = sqrtT*self.mean[k]
 
-    def thermal_loss(self, T, nth, k):
+    def thermal_loss(self, T, nbar, k):
         r""" Implements the thermal loss channel in mode k by amplitude loss amount \sqrt{T}
         unlike the loss channel, here the ancilliary mode that goes into the second arm of the
         beam splitter is prepared in a thermal state with mean photon number nth """
```
Comment (co9olguy) [inline strawberryfields/backends/gaussianbackend/gaussiancircuit.py]: I cleaned up this docstring in `master`. Let's remember to double check that the merge doesn't revert back to the old one

File: doc/conventions/others.rst (hunk 1 above)
Comment (josh146) [reply doc/conventions/others.rst]: \a is already defined as :math:`\hat{a}` :)

File: strawberryfields/ops.py (hunk 5 above)
Comment (josh146) [reply strawberryfields/ops.py]: I assume so? This is one of Ville's original comments

**Pull request merged**